File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        // finding pivot
        int n = nums.length, low = 0, high = n - 1, mid = (n - 1) / 2;
        if(nums[low] < nums[high])
            mid = -1;
        else
            while(low < high) {
                mid = (high + low) / 2;
                if(nums[mid] > nums[low])
                    low = mid;
                else
                    high = mid;
            }
        int start = mid + 1;
        // finding target
        int temp; low = 0; high = n - 1;
        while(low <= high) {
            mid = (high + low) / 2;
            temp = (start + mid) % n;
            if(nums[temp] == target)
                return temp;
            else if(nums[temp] < target)
                low = mid + 1;
            else
                high = mid - 1;
        }
        return -1;
    }
}

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        n = len(nums)
        high = n - 1
        low = 0
        while low <= high:
            mid = (low + high) // 2
            if nums[low] > nums[mid]:
                if nums[mid-1] < nums[mid]:
                    high = mid - 1
                else:
                    rot = mid
                    break
            elif nums[mid] > nums[high]:
                if nums[mid+1] > nums[mid]:
                    low = mid + 1
                else:
                    rot = mid + 1
                    break
            else:
                rot = low
                break
        bs = lambda x: (x + rot) % n
        high = n - 1
        low = 0
        while low <= high:
            mid = (low + high) // 2
            if nums[bs(mid)] == target:
                return bs(mid)
            elif nums[bs(mid)] > target:
                high = mid - 1
            else:
                low = mid + 1
        return -1